File: sem4/jpp/list2/cpp/testGF1234577Field.cpp
#include "GF1234577Field.hpp"

#include <cassert>
#include <iostream>
#include <sstream>

const int CHARACTERISTIC = 1234577;

// Initialization
void testInitialization() {
    GF1234577Field field = GF1234577Field(10);
    assert(field.getValue() == 10);
}

void testInitializationWithNegativeNumber() {
    GF1234577Field field = GF1234577Field(-10);
    assert(field.getValue() == CHARACTERISTIC - 10);
}

// Comparison operators
void testEqualOperatorTrue() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(10);

    bool result = (field1 == field2);
    assert(result == true);
}

void testEqualOperatorFalse() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    bool result = (field1 == field2);
    assert(result == false);
}

void testNotEqualOperatorTrue() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    bool result = (field1 != field2);
    assert(result == true);
}

void testNotEqualOperatorFalse() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 != field2);
    assert(result == false);
}

void testLessThanOperatorTrue() {
    GF1234577Field field1 = GF1234577Field(5);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 < field2);
    assert(result == true);
}

void testLessThanOperatorFalse() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    bool result = (field1 < field2);
    assert(result == false);
}

void testLessThanOrEqualOperatorTrue() {
    GF1234577Field field1 = GF1234577Field(5);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 <= field2);
    assert(result == true);
}

void testLessThanOrEqualOperatorTrueEqual() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 <= field2);
    assert(result == true);
}

void testLessThanOrEqualOperatorFalse() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    bool result = (field1 <= field2);
    assert(result == false);
}

void testGreaterThanOperatorTrue() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    bool result = (field1 > field2);
    assert(result == true);
}

void testGreaterThanOperatorFalse() {
    GF1234577Field field1 = GF1234577Field(5);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 > field2);
    assert(result == false);
}

void testGreaterThanOrEqualOperatorTrue() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    bool result = (field1 >= field2);
    assert(result == true);
}

void testGreaterThanOrEqualOperatorTrueEqual() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 >= field2);
    assert(result == true);
}

void testGreaterThanOrEqualOperatorFalse() {
    GF1234577Field field1 = GF1234577Field(5);
    GF1234577Field field2 = GF1234577Field(10);
    bool result = (field1 >= field2);
    assert(result == false);
}

// Arithmetic operators
void testAdditionOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    GF1234577Field result = field1 + field2;
    assert(result.getValue() == 15);
}

void testSubtractionOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    GF1234577Field result = field1 - field2;
    assert(result.getValue() == 5);
}

void testMultiplicationOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    GF1234577Field result = field1 * field2;
    assert(result.getValue() == 50);
}

void testDivisionOperator() {
    GF1234577Field field1 = GF1234577Field(1);
    GF1234577Field field2 = GF1234577Field(2);
    GF1234577Field result = field1 / field2;
    assert(result.getValue() == 617289);
}

// Assignment operators
void testAssignmentOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    field1 = field2;
    assert(field1.getValue() == 5);
}

void testAssignmentAdditionOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    field1 += field2;
    assert(field1.getValue() == 15);
}

void testAssignmentSubtractionOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    field1 -= field2;
    assert(field1.getValue() == 5);
}

void testAssignmentSubtractionOperatorWithNegativeNumber() {
    GF1234577Field field1 = GF1234577Field(5);
    GF1234577Field field2 = GF1234577Field(10);
    field1 -= field2;
    assert(field1.getValue() == CHARACTERISTIC - 5);
}

void testAssignmentMultiplicationOperator() {
    GF1234577Field field1 = GF1234577Field(10);
    GF1234577Field field2 = GF1234577Field(5);
    field1 *= field2;
    assert(field1.getValue() == 50);
}

void testAssignmentDivisionOperator() {
    GF1234577Field field1 = GF1234577Field(1);
    GF1234577Field field2 = GF1234577Field(2);
    field1 /= field2;
    assert(field1.getValue() == 617289);
}

// Output stream operator
void testOutputStreamOperator() {
    GF1234577Field field = GF1234577Field(10);
    std::stringstream out;
    out << field;
    assert(out.str() == "10");
}

// Conversion operators
void testIntOperator() {
    GF1234577Field field = GF1234577Field(10);
    assert(int(field) == 10);
}

void testFloatOperator() {
    GF1234577Field field = GF1234577Field(10);
    assert(float(field) == 10.0f);
}

void testDoubleOperator() {
    GF1234577Field field = GF1234577Field(10);
    assert(double(field) == 10.0);
}


int main() {
    testInitialization();
    testInitializationWithNegativeNumber();

    testEqualOperatorTrue();
    testEqualOperatorFalse();
    testNotEqualOperatorTrue();
    testNotEqualOperatorFalse();
    testLessThanOperatorTrue();
    testLessThanOperatorFalse();
    testLessThanOrEqualOperatorTrue();
    testLessThanOrEqualOperatorTrueEqual();
    testLessThanOrEqualOperatorFalse();
    testGreaterThanOperatorTrue();
    testGreaterThanOperatorFalse();
    testGreaterThanOrEqualOperatorTrue();
    testGreaterThanOrEqualOperatorTrueEqual();
    testGreaterThanOrEqualOperatorFalse();

    testAdditionOperator();
    testSubtractionOperator();
    testMultiplicationOperator();
    testDivisionOperator();

    testAssignmentOperator();
    testAssignmentAdditionOperator();
    testAssignmentSubtractionOperator();
    testAssignmentSubtractionOperatorWithNegativeNumber();
    testAssignmentMultiplicationOperator();
    testAssignmentDivisionOperator();

    testOutputStreamOperator();

    testIntOperator();
    testFloatOperator();
    testDoubleOperator();

    std::cout << "All tests passed!\n"; 
}

File: sem4/jpp/list2/python/test_gf1234577field.py
import pytest
from gf1234577field import GF1234577Field

CHARACTERISTIC = 1234577


class TestGF1234577Field:
    # Initialization
    def test_initialization(self):
        field = GF1234577Field(10)
        assert field.value == 10

    def test_initialization_with_negative_number(self):
        field = GF1234577Field(-10)
        assert field.value == CHARACTERISTIC - 10

    # Comparison operators
    def test_equal_operator_true(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(10)
        result = field1 == field2
        assert result == True

    def test_equal_operator_false(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 == field2
        assert result == False

    def test_not_equal_operator_true(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 != field2
        assert result == True

    def test_not_equal_operator_false(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(10)
        result = field1 != field2
        assert result == False

    def test_less_than_operator_true(self):
        field1 = GF1234577Field(5)
        field2 = GF1234577Field(10)
        result = field1 < field2
        assert result == True

    def test_less_than_operator_false(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 < field2
        assert result == False

    def test_less_than_or_equal_operator_true(self):
        field1 = GF1234577Field(5)
        field2 = GF1234577Field(10)
        result = field1 <= field2
        assert result == True

    def test_less_than_or_equal_operator_true_equal(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(10)
        result = field1 <= field2
        assert result == True

    def test_less_than_or_equal_operator_false(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 <= field2
        assert result == False

    def test_greater_than_operator_true(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 > field2
        assert result == True

    def test_greater_than_operator_false(self):
        field1 = GF1234577Field(5)
        field2 = GF1234577Field(10)
        result = field1 > field2
        assert result == False

    def test_greater_than_or_equal_operator_true(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 >= field2
        assert result == True

    def test_greater_than_or_equal_operator_true_equal(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(10)
        result = field1 >= field2
        assert result == True

    def test_greater_than_or_equal_operator_false(self):
        field1 = GF1234577Field(5)
        field2 = GF1234577Field(10)
        result = field1 >= field2
        assert result == False

    # Arithmetic operators
    def test_addition_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 + field2
        assert result.value == 15

    def test_subtraction_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 - field2
        assert result.value == 5

    def test_multiplication_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        result = field1 * field2
        assert result.value == 50

    def test_division_operator(self):
        field1 = GF1234577Field(1)
        field2 = GF1234577Field(2)
        result = field1 / field2
        assert result.value == 617289

    # Assignment operators
    def test_assignment_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        field1 = field2
        assert field1.value == 5

    def test_assignment_addition_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        field1 += field2
        assert field1.value == 15

    def test_assignment_subtraction_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        field1 -= field2
        assert field1.value == 5

    def test_assignment_subtraction_operator_with_negative_number(self):
        field1 = GF1234577Field(5)
        field2 = GF1234577Field(10)
        field1 -= field2
        assert field1.value == CHARACTERISTIC - 5

    def test_assignment_multiplication_operator(self):
        field1 = GF1234577Field(10)
        field2 = GF1234577Field(5)
        field1 *= field2
        assert field1.value == 50

    def test_assignment_division_operator(self):
        field1 = GF1234577Field(1)
        field2 = GF1234577Field(2)
        field1 /= field2
        assert field1.value == 617289

    # Conversion operators
    def test_str_operator(self):
        field = GF1234577Field(10)
        assert str(field) == "10"

    def test_repr_operator(self):
        field = GF1234577Field(10)
        assert repr(field) == "GF1234577Field(10)"

    def test_int_operator(self):
        field = GF1234577Field(10)
        assert int(field) == 10

    def test_float_operator(self):
        field = GF1234577Field(10)
        assert float(field) == 10.0
